hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestPositionBasedMergingFallback.scala

```diff
@@ -0,0 +1,192 @@
+      option(HoodieReaderConfig.FILE_GROUP_READER_ENABLED.key(), "true").
+      option(HoodieWriteConfig.WRITE_RECORD_POSITIONS.key(), updateWithRecordPositions).
+      mode(Append).
+      save(basePath)
+
+    val deletesData = Seq((10, "4", "D",  true), (10, "3", "C", true))
+
+    val deletes = sparkSession.createDataFrame(deletesData).toDF(columns: _*)
+    deletes.write.format("hudi").
+      option(RECORDKEY_FIELD.key(), "key").
+      option(PRECOMBINE_FIELD.key(), "ts").
+      option("hoodie.table.name", "test_table").
+      option(TABLE_TYPE.key(), "MERGE_ON_READ").
+      option(OPERATION.key(), "upsert").
+      option(HoodieStorageConfig.LOGFILE_DATA_BLOCK_FORMAT.key(), "parquet").
+      option(DataSourceWriteOptions.RECORD_MERGER_STRATEGY.key(), HoodieRecordMerger.DEFAULT_MERGER_STRATEGY_UUID).
+      option(DataSourceWriteOptions.RECORD_MERGER_IMPLS.key(), "org.apache.hudi.HoodieSparkRecordMerger").
+      option(HoodieReaderConfig.FILE_GROUP_READER_ENABLED.key(), "true").
+      option(HoodieWriteConfig.WRITE_RECORD_POSITIONS.key(), deleteWithRecordPositions).
+      mode(Append).
+      save(basePath)
+
+
+    val secondUpdateData = Seq((14, "5", "E_3", false), (3, "3", "C_3", false))
+    val secondUpdates = sparkSession.createDataFrame(secondUpdateData).toDF(columns: _*)
+    secondUpdates.write.format("hudi").
+      option(RECORDKEY_FIELD.key(), "key").
+      option(PRECOMBINE_FIELD.key(), "ts").
+      option("hoodie.table.name", "test_table").
+      option(TABLE_TYPE.key(), "MERGE_ON_READ").
+      option(OPERATION.key(), "upsert").
+      option(HoodieStorageConfig.LOGFILE_DATA_BLOCK_FORMAT.key(), "parquet").
+      option(DataSourceWriteOptions.RECORD_MERGER_STRATEGY.key(), HoodieRecordMerger.DEFAULT_MERGER_STRATEGY_UUID).
+      option(DataSourceWriteOptions.RECORD_MERGER_IMPLS.key(), "org.apache.hudi.HoodieSparkRecordMerger").
+      option(HoodieReaderConfig.FILE_GROUP_READER_ENABLED.key(), "true").
+      option(HoodieWriteConfig.WRITE_RECORD_POSITIONS.key(), secondUpdateWithPositions).
+      mode(Append).
+      save(basePath)
+
+    val df = sparkSession.read.format("hudi").
+      option(DataSourceWriteOptions.RECORD_MERGER_STRATEGY.key(), HoodieRecordMerger.DEFAULT_MERGER_STRATEGY_UUID).
+      option(DataSourceWriteOptions.RECORD_MERGER_IMPLS.key(), "org.apache.hudi.HoodieSparkRecordMerger").
+      option(HoodieReaderConfig.FILE_GROUP_READER_ENABLED.key(), "true").
+      option(HoodieReaderConfig.MERGE_USE_RECORD_POSITIONS.key(), "true").load(basePath)
+    val finalDf = df.select("ts", "key", "name")
+    val finalColumns = Seq("ts", "key", "name")
+
+    val finalExpectedData = Seq(
+      (11, "1", "A_1"),
+      (10, "2", "B"),
+      (14, "5", "E_3"))
+
+    val expectedDf = sparkSession.createDataFrame(finalExpectedData).toDF(finalColumns: _*)
+
+    assertEquals(0, finalDf.except(expectedDf).count())
+    assertEquals(0, expectedDf.except(finalDf).count())
```

Comment:
this is cool but do we also test logic under `needToDoHybridStrategy` anywhere? Basically, the scenario "if it's a delete record and the key is null, then we need to still use positions".

Reply:
needToDoHybridStrategy is needed when we just read the positions from the delete block using the roaring bitmap. The bitmap only stores the positions, so even if the recordkey is not null in the record, we still don't have that information.